config/v1/types_dns.go

```diff
@@ -26,6 +26,25 @@ type DNSSpec struct {
 	// For example, given the base domain `openshift.example.com`, an API server
 	// DNS record may be created for `cluster-api.openshift.example.com`.
 	BaseDomain string `json:"baseDomain"`
+	// publicZone is the location where all the DNS records that are publicly accessible to
+	// the internet exist.
+	// If this field is nil, no public records should be created.
+	// +optional
+	PublicZone *DNSZone `json:"publicZone,omitempty"`
+	// privateZone is the location where all the DNS records that are only available internally
+	// to the cluster exist.
+	PrivateZone DNSZone `json:"privateZone"`
+}
+
+// DNSZone is used to define a DNS hosted zone.
+// A zone can be identified by an ID or tags.
+type DNSZone struct {
```

Comment:
So is this meant to be a platform specific, opaque struct that will be able to address the zone on any supported platform? Will ID+Tags cover all those or do we need a union type that allows for a more strongly typed config object per supported platform?

Reply:
> Will ID+Tags
ID or Tags, I don't think you need both...

Reply:
I think ID should cover it. @openshift/sig-network-edge WDYT?

Reply:
Either should work for us.  What's more convenient for the user? Are we going to respond to updates to this config?

Reply:
Is an ID string AWS-specific or cross-platform?  Will other providers need additional information to discover/update the zones?

Reply:
>Is an ID string AWS-specific or cross-platform? Will other providers need additional information to discover/update the zones?

I was trying to ask this but perhaps in an unclear way... it seems like this value becomes an opaque platform-specific value that would require the consumer to use the `Infrastructure.Status.Platform` to infer how to decode...

Reply:
> ... it seems like this value becomes an opaque platform-specific value that would require the consumer to use the Infrastructure.Status.Platform to infer how to decode...

Are we supporting cross-platform DNS providers (e.g. an AWS cluster with Route 53 for internal DNS and... some other DNS provider for the public zone)?  I think AWS load balancers may require Route 53?  If we aren't worried about per-platform zone entries, then we can get the platform information from [the infrastructure config][1].

While a string might be sufficient for all platforms, using:

```go
type DNSZone struct {
  ID *string `json:"id"`
}
```

gives us room to add more properties if we end up needing them later.

[1]: https://github.com/openshift/api/blob/21e40de482627b2047cbf430594540413539a442/config/v1/types_infrastructure.go#L37

Reply:
> ... so id is not known when `DNS` object is created...

Ah, that makes sense.  I guess the ingress operator could perform the tag -> ID lookup in its first round and then push its own update to the config to get more efficient lookups later, if we're concerned about API quotas.

Reply:
So we need an ID for the public zone and tags for the private zone? If so it seems like AWS implies details leaking into this API, but maybe I'm misunderstanding?

If tags are sufficient to address the private zone, should the be sufficient to address the public zone?